framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java

```diff
@@ -0,0 +1,584 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Red Hat, Inc. Distributed under license by Red Hat, Inc.
+ * All rights reserved. This program is made available under the terms of the
+ * Eclipse Public License v1.0 which accompanies this distribution, and is
+ * available at http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors: Red Hat, Inc.
+ ******************************************************************************/
+package org.jboss.tools.rsp.server.filewatcher;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.jboss.tools.rsp.server.spi.filewatcher.FileWatcherEvent;
+import org.jboss.tools.rsp.server.spi.filewatcher.IFileWatcherEventListener;
+import org.jboss.tools.rsp.server.spi.filewatcher.IFileWatcherService;
+
+public class FileWatcherService implements Runnable, IFileWatcherService {
+	// The watch service from java.nio
+	private WatchService watchService;
+	
+	// A thread which runs and listens for events
+	private Thread serviceThread;
+
+	/* 
+	 *  A map of paths and associated listeners that have been specifically
+	 *  requested by some client to be listened to.
+	 *  
+	 */ 
+	private HashMap<Path, List<RegistrationRequest>> requests = new HashMap<>();
+	
+	/*
+	 * A map of path -> watch key, listing each and every folder 
+	 * that we are subscribed to receive events from. 
+	 * This map will contain much more data than the requests map. 
+	 * It will possibly contain recursive subscriptions, as well 
+	 * as parent subscriptions up to the root of the filesystem. 
+	 */
+	private HashMap<Path, WatchKey> subscriptions = new HashMap<>();
+	
+	private boolean closing = false;
+
+	public FileWatcherService() {
+	}
+	
+	private void log(Exception e) {
+		// TODO log
+		e.printStackTrace();
+	}
+
+	private synchronized void setClosing(boolean b) {
+		this.closing = b;
+	}
+	
+	private synchronized boolean isClosing() {
+		return this.closing;
+	}
+	
+	public void start() throws IllegalStateException {
+		try {
+			watchService = FileSystems.getDefault().newWatchService();
+		} catch (IOException e) {
+			log(e);
+		}
+		if( watchService == null )
+			throw new IllegalStateException();
+		serviceThread = new Thread(this);
```

Comment:
I'd favor a more modern pattern where you submit ```this``` to an ExecutorService. The benefit is that the service can be forcefully shut down. 

See here in AbstractPoller: https://github.com/redhat-developer/rsp-server/blob/master/framework/bundles/org.jboss.tools.rsp.server.spi/src/main/java/org/jboss/tools/rsp/server/spi/model/polling/AbstractPoller.java#L40